fix(#2108): dropdown width related issues

syedszeeshan opened the pull request:
# Before (the change)

# After (the change)


## Make sure that you've checked the boxes below before you submit the PR

- [x] I have read and followed the [setup steps](https://goa-dio.atlassian.net/wiki/spaces/DS/pages/3086385489/Contribution+-+web+content#%5BinlineExtension%5DCode-contributions)
- [ ] I have created necessary unit tests
- [x] I have tested the functionality in both React and Angular.

## Steps needed to test

Playground Test Code

Width specified
```
export default function Dropdown() {
  return (
    <GoAFormItem label="Dropdown" requirement="optional">
      <GoADropdown
        name="dropdown"
        width="30ch"
        placeholder="Select a value"
        onChange={function (name: string, values: string[] | string): void {
          throw new Error("Function not implemented. ");
        }}
      >
        <GoADropdownItem key="_none_" name="none" value="" label="None" />
        <GoADropdownItem key="_some_" name="some" value="" label="Some Long Value Some Long Value" />
      </GoADropdown>
    </GoAFormItem>
  );
}
```

No width specified
```
export default function Dropdown() {
  return (
    <GoAFormItem label="Dropdown" requirement="optional">
      <GoADropdown
        name="dropdown"
        placeholder="Select a value"
        onChange={function (name: string, values: string[] | string): void {
          throw new Error("Function not implemented. ");
        }}
      >
        <GoADropdownItem key="_none_" name="none" value="" label="None" />
        <GoADropdownItem key="_some_" name="some" value="" label="Some Long Value Some Long Value" />
      </GoADropdown>
    </GoAFormItem>
  );
}
```

lizhuomeng71 commented:
The popover's maximum width remains fixed at 320px, even when the dropdown width is set to a percentage value, leading to inconsistent sizing.

Please see this example
```

    <GoAFormItem label="Dropdown" requirement="optional">
      <GoADropdown
        name="dropdown"
        width="50%"
        placeholder="Select a value"
        onChange={function (name: string, values: string[] | string): void {
          throw new Error("Function not implemented. ");
        }}
      >
        <GoADropdownItem key="_none_" name="none" value="" label="None" />
        <GoADropdownItem key="_some_" name="some" value="" label="Some Long Value Some Long Value" />
      </GoADropdown>
    </GoAFormItem>
```

syedszeeshan commented:
> The popover's maximum width remains fixed at 320px, even when the dropdown width is set to a percentage value, leading to inconsistent sizing.
> 
> Please see this example
> 
> ```
> 
>     <GoAFormItem label="Dropdown" requirement="optional">
>       <GoADropdown
>         name="dropdown"
>         width="50%"
>         placeholder="Select a value"
>         onChange={function (name: string, values: string[] | string): void {
>           throw new Error("Function not implemented. ");
>         }}
>       >
>         <GoADropdownItem key="_none_" name="none" value="" label="None" />
>         <GoADropdownItem key="_some_" name="some" value="" label="Some Long Value Some Long Value" />
>       </GoADropdown>
>     </GoAFormItem>
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/7528736/375561672-295a2263-9c04-4168-8227-8fb3d5446166.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2Njc1MTIsIm5iZiI6MTcyODY2NzIxMiwicGF0aCI6Ii83NTI4NzM2LzM3NTU2MTY3Mi0yOTVhMjI2My05YzA0LTQxNjgtODIyNy04ZmIzZDU0NDYxNjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTFUMTcyMDEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODI0MTVlZDIwZDU3YzgxMTFlZmU1NDMwYjBkYzRjYTc1M2VkMmM3NDNkNDgzN2M2ZTA3MGY0ZTg1ZWNlYmRjMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T1URAl5sUwsoQp4dL_AqnccoVCtbHwE0Zr6dJ2c0BOQ)

@lizhuomeng71 Thanks, I have fixed it and updated teh PR, ready for testing.

syedszeeshan commented:
> The popover's maximum width remains fixed at 320px, even when the dropdown width is set to a percentage value, leading to inconsistent sizing.
> 
> Please see this example
> 
> ```
> 
>     <GoAFormItem label="Dropdown" requirement="optional">
>       <GoADropdown
>         name="dropdown"
>         width="50%"
>         placeholder="Select a value"
>         onChange={function (name: string, values: string[] | string): void {
>           throw new Error("Function not implemented. ");
>         }}
>       >
>         <GoADropdownItem key="_none_" name="none" value="" label="None" />
>         <GoADropdownItem key="_some_" name="some" value="" label="Some Long Value Some Long Value" />
>       </GoADropdown>
>     </GoAFormItem>
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/7528736/375561672-295a2263-9c04-4168-8227-8fb3d5446166.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2Njc3MzAsIm5iZiI6MTcyODY2NzQzMCwicGF0aCI6Ii83NTI4NzM2LzM3NTU2MTY3Mi0yOTVhMjI2My05YzA0LTQxNjgtODIyNy04ZmIzZDU0NDYxNjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTFUMTcyMzUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDc2M2Y5MTk4YjMwYWRmZGM5ZDY2ZDRlZjhhOGQzZWQxNjQ4OTBhMjUzN2ZkMDkzNjgzYTY2ZGVkZmRlMWFkNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.upxs4sfbeeXfedo7RgjPPLYHfu2yppBmS2Gm_LXo_Bk)

@lizhuomeng71 Thanks, it is fixed and PR has been updated. Ready for testing.

lizhuomeng71 commented:
This issue is verified on react and angular

chrisolsen commented on libs/react-components/src/lib/dropdown/dropdown.tsx:
```diff
@@ -102,7 +104,7 @@ export function GoADropdown(props: GoADropdownProps): JSX.Element {
       placeholder={props.placeholder}
       testid={props.testId}
-      width={props.width}
+      width={props.width || (!hasDropdownItems ? "100%" : undefined)}
```
Is this logic needed here? Ideally all logic should exist within the web-component.

syedszeeshan replied:
PR has been updated.

chrisolsen commented on libs/web-components/src/components/dropdown/Dropdown.svelte:
```diff
@@ -94,6 +112,11 @@
     _eventHandler = _filterable
       ? new ComboboxKeyUpHandler(_inputEl)
       : new DropdownKeyUpHandler(_inputEl);
+
+    // Immediately set the width if provided
+    if (width) {
+      _width = width;
+    }
```
Since the `_width` has been made to be reactive, I don't think this block of code is required.

syedszeeshan replied:
PR Updated

syedszeeshan replied:
PR has been updated.

chrisolsen commented on libs/web-components/src/components/dropdown/Dropdown.svelte:
```diff
@@ -173,7 +195,7 @@
   function bind() {
     syncFilteredOptions();
-    if (!width) {
+    if (!width && _options.length > 0) {
```
`bind` is only called after an event, containing an option, has been received, so this additional check is not needed.

syedszeeshan replied:
PR Updated

syedszeeshan replied:
PR has been updated.

chrisolsen commented on libs/web-components/src/components/dropdown/Dropdown.svelte:
```diff
@@ -572,6 +595,7 @@
   style={`
       ${calculateMargin(mt, mr, mb, ml)};
       --width: ${_width};
+      width: ${_width};
```
The setting of the style width is already being done within the styles via the `--width` css value being set, but is also being set based on the screen size via the `@media` rule.

syedszeeshan replied:
PR has been updated.

chrisolsen commented on libs/web-components/src/components/dropdown/Dropdown.svelte:
```diff
@@ -600,7 +624,10 @@
       data-testid="option-list"
-      width={`${_popoverMaxWidth || 100}px`}
+      width={`${_popoverMaxWidth || 300}px`}
+      maxwidth={_popoverMaxWidth
+        ? `${_popoverMaxWidth}px`
+        : getLongestChildWidth(_options) || undefined}
```
Does it make sense to set a width and a max-width?

syedszeeshan replied:
If we dont set the maxwidth from within the dropdown component, the popover width can extend up to a maximum of 320px.

syedszeeshan replied:
Not setting the maxwidth from the dropdown component anymore. Rather, updated the logic in the Popover componentm, as discussed.

syedszeeshan commented on libs/web-components/src/components/popover/Popover.svelte:
```diff
@@ -101,6 +107,23 @@
+    } else {
+      return width; // For non-px units use the width as maxwidth
+    }
+  }
```
Replaced with this `style("max-width", width ? `max(${width}, ${maxwidth})` : maxwidth),`

syedszeeshan commented on libs/web-components/src/components/popover/Popover.svelte:
```diff
@@ -68,13 +72,15 @@
   let _focusTrapEl: HTMLElement;
   let _initFocusedEl: HTMLElement;
   let _sectionHeight: number;
+  let _effectiveMaxWidth: string;
```
replaced with `style("max-width", width ? `max(${width}, ${maxwidth})` : maxwidth),`

chrisolsen commented on libs/web-components/src/components/popover/Popover.svelte:
```diff
@@ -100,8 +104,6 @@
       ) as HTMLElement) || _targetEl;
   });
 
-  // Functions
-
```
Keep this in. It serves as a reminder for others to keep the code a little orgnanized.

syedszeeshan replied:
PR updated.

lizhuomeng71 commented:
this issue is verified on react and angular

tzuge commented:
:tada: This PR is included in version 1.17.0-alpha.114 :tada:

The release is available on:
- [npm package (@alpha dist-tag)](https://www.npmjs.com/package/@abgov/web-components/v/1.17.0-alpha.114)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/web-components-v1.17.0-alpha.114)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tzuge commented:
:tada: This PR is included in version 5.0.0-alpha.10 :tada:

The release is available on:
- [npm package (@alpha dist-tag)](https://www.npmjs.com/package/@abgov/react-components/v/5.0.0-alpha.10)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/react-v5.0.0-alpha.10)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tzuge commented:
:tada: This PR is included in version 1.28.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@abgov/web-components/v/1.28.0)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/web-components-v1.28.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tzuge commented:
:tada: This PR is included in version 5.3.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@abgov/react-components/v/5.3.0)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/react-v5.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tzuge commented:
:tada: This PR is included in version 3.0.0-alpha.9 :tada:

The release is available on:
- [npm package (@alpha dist-tag)](https://www.npmjs.com/package/@abgov/angular-components/v/3.0.0-alpha.9)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/angular-v3.0.0-alpha.9)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tzuge commented:
:tada: This PR is included in version 3.2.1 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@abgov/angular-components/v/3.2.1)
- [GitHub release](https://github.com/GovAlta/ui-components/releases/tag/angular-v3.2.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
